frontend/device/generic/device.lua

```diff
@@ -142,7 +142,12 @@ function Device:onPowerEvent(ev)
         self.screen_saver_mode = true
         UIManager:scheduleIn(0.1, function()
           local network_manager = require("ui/network/manager")
-          if network_manager.wifi_was_on then
+          -- NOTE: wifi_was_on does not necessarily mean that WiFi is *currently* on! It means *we* enabled it.
+          --       This is critical on Kobos (c.f., #3936), where it might still be on from KSM or Nickel,
+          --       without us being aware of it (i.e., wifi_was_on still unset or false),
+          --       because suspend will at best fail, and at worst deadlock the system if WiFi is on,
+          --       regardless of who enabled it!
+          if network_manager.wifi_was_on or network_manager:isWifiOn() then
```

Comment:
I'm hoping this doesn't behave stupidly on devices where isWifiOn() isn't set?

Reply:
Because that's basically everywhere except Android and Kobo, so that would be... not good? :D.

Reply:
It'd just return nil, which is false-y for the sake of the or.

Reply:
Regarding the comment, did someone change the behavior where we *always* try turn off frontlight and wifi and all those things that can fail suspend regardless of their status or something? Because if so that was all in there on purpose.

Reply:
FL is properly turned off in the device-specific BeforeSuspend, and as for WiFi, AFAICT, that's the bit that made the decision of whether or not to kill WiFi, except the check is not accurate, because wifi_was_on isn't.

Reply:
Meaning that from what I gathered, in the current state of the codebase, while we were always shutting down the FL, we weren't doing so for WiFi, instead relying on this flawed test.

I can't say if that was always the case, though ;).